### PR TITLE
vs/Automate Firefox candidate URL resolution

### PR DESCRIPTION
**Pull request by vsangereanMOZ**
### Relevant Links

Bugzilla: N/A
TestRail: N/A

### Description of Code / Doc Changes

While working on my Firefox train branches and having to manually run the ESR/Dot tests, I saw that the links built in archive.mozilla are the same. Only the build nr changes. 
I made this quick solution so that the manual team can start the tests by only entering the build number, and the Starfox team to just verify the results. 

I kept the old logic available just so this can be overwritten or bypassed fast if something changes. 

Thank you!

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -89,6 +98,19 @@ jobs:
         with:
           python-version: "3.13"
 
+      - name: Resolve Windows candidate URL
+        if: ${{ github.event_name == 'workflow_dispatch' }}
+        shell: pwsh
+        env:
+          FIREFOX_VERSION: ${{ inputs.firefox_version }}
+          INSTALLER_OVERRIDE: ${{ inputs.win_installer_link }}
```
[the remark this answers is not in the document]
Comment (vsangereanMOZ) [reply .github/workflows/main.yml:106]: The download step runs only after the resolver succeeds. The resolver either sets MANUAL_DOWNLOAD_LINK or exits with an error, which prevents later steps from running. Therefore, the value cannot be empty on this path. Also, the suggested guard would fall into the else block and call Invoke-WebRequest with an empty URL, so it would not fail more gracefully.

**Review by ben-c-at-moz — CHANGES_REQUESTED** [review]
We considered building urls from version numbers previously, but because we don't always know which types of Fx executables we'll be looking at, we haven't attempted it. This solution is far too verbose and would be difficult to maintain, and doesn't account for the edge cases we were expecting when the original scripts were written. The choice to use `unittest` to test the solution is also strange given how much PyTest we're using in the repo.

**Comment by vsangereanMOZ** [comment]
> We considered building urls from version numbers previously, but because we don't always know which types of Fx executables we'll be looking at, we haven't attempted it. This solution is far too verbose and would be difficult to maintain, and doesn't account for the edge cases we were expecting when the original scripts were written. The choice to use `unittest` to test the solution is also strange given how much PyTest we're using in the repo.

I made the pytest changes. 
I think this approach is still useful for our manual team because they normally work with official en-US candidate builds.
They can enter the Firefox version once instead of finding and copying three platform-specific URLs, and the resolver automatically handles cases where the candidate is build2 or later. 

The existing URL fields are still available as overrides, so we are not removing the current workflow or support for unusual artifacts. The resolver only provides a simpler path for the common case, with the logic centralized and tested rather than duplicated in the workflow.

This will save us some times each week.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -336,8 +368,6 @@ jobs:
 
       - name: Install dependencies
```
[the remark this answers is not in the document]
Comment (vsangereanMOZ) [reply .github/workflows/main.yml:342]: I checked this against collect_executables.py. When called with -n, it first returns the existing FX_VERSION environment value

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -482,8 +523,6 @@ jobs:
...
```
[the remark this answers is not in the document]
Comment (vsangereanMOZ) [reply .github/workflows/main.yml:488]: I checked this against collect_executables.py. When called with -n, it first returns the existing FX_VERSION environment value

**Review by ben-c-at-moz — CHANGES_REQUESTED** [review]
I still don't really think the direction of this PR is optimal. I would prefer a new flow that calls main via workflow_dispatch, rather than changing main.yml. I still think the script that generates the URL is too verbose/complex, and I don't think we need this full a test suite for it (though maybe this is a kick in the pants to get tests for our scripts).

File: .github/workflows/run-firefox-candidate.yml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            echo "Invalid Firefox candidate version: $FIREFOX_VERSION" >&2
+            echo "Expected examples: 155.0.1, 153.0esr, or 156.0b2" >&2
+            exit 1
+          fi
+
+          candidate_root="https://archive.mozilla.org/pub/firefox/candidates/${FIREFOX_VERSION}-candidates"
+          listing="$(curl --fail --location --silent --show-error --max-time 30 "$candidate_root/")"
+          # Archive listings expose candidate directories as /buildN/ links.
+          build="$(
+            grep -oE '/build[1-9][0-9]*/' <<< "$listing" |
```
[the remark this answers is not in the document]
Comment (vsangereanMOZ) [reply .github/workflows/run-firefox-candidate.yml]: not chaning. this should be sufficient.

File: .github/workflows/run-firefox-candidate.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+          mac_url="$base_url/mac/en-US/Firefox%20${FIREFOX_VERSION}.dmg"
+          linux_url="$base_url/linux-x86_64/en-US/firefox-${FIREFOX_VERSION}.tar.xz"
+
+          for url in "$win_url" "$mac_url" "$linux_url"; do
+            curl --fail --location --silent --show-error --head --max-time 30 "$url" > /dev/null
+          done
+
+          echo "Resolved Firefox $FIREFOX_VERSION build$build"
+          gh workflow run main.yml \
+            --ref "$GITHUB_REF_NAME" \
```
[the remark this answers is not in the document]
Comment (vsangereanMOZ) [reply .github/workflows/run-firefox-candidate.yml:65]: not in scope.

File: .github/workflows/run-firefox-candidate.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+          win_url="$base_url/win64/en-US/Firefox%20Setup%20${FIREFOX_VERSION}.exe"
+          mac_url="$base_url/mac/en-US/Firefox%20${FIREFOX_VERSION}.dmg"
+          linux_url="$base_url/linux-x86_64/en-US/firefox-${FIREFOX_VERSION}.tar.xz"
+
+          for url in "$win_url" "$mac_url" "$linux_url"; do
+            curl --fail --location --silent --show-error --head --max-time 30 "$url" > /dev/null
+          done
+
+          echo "Resolved Firefox $FIREFOX_VERSION build$build"
+          gh workflow run main.yml \
```
[the remark this answers is not in the document]
Comment (ben-c-at-moz) [reply .github/workflows/run-firefox-candidate.yml:70]: This is correct. Make sure it's clear in the instructions that this is only for beta, or make sure that other logic in the workflow will correctly ascertain the channel based on context.

[the remark this answers is not in the document]
Comment (ben-c-at-moz) [reply .github/workflows/run-firefox-candidate.yml:70]: I believe `"rc"` is also a valid value for `FX_CHANNEL`.

File: .github/workflows/run-firefox-candidate.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            exit 1
+          fi
+
+          base_url="$candidate_root/build$build"
+          win_url="$base_url/win64/en-US/Firefox%20Setup%20${FIREFOX_VERSION}.exe"
+          mac_url="$base_url/mac/en-US/Firefox%20${FIREFOX_VERSION}.dmg"
+          linux_url="$base_url/linux-x86_64/en-US/firefox-${FIREFOX_VERSION}.tar.xz"
+
+          for url in "$win_url" "$mac_url" "$linux_url"; do
+            curl --fail --location --silent --show-error --head --max-time 30 "$url" > /dev/null
```
Comment (ben-c-at-moz) [inline .github/workflows/run-firefox-candidate.yml]: I think you should shorten this max time. The response to a HEAD request shouldn't nearly take that long.

**Review by ben-c-at-moz — DISMISSED** [review]
This is better, thanks. I do have some comments though

**Review by ben-c-at-moz — APPROVED** [review]
Let's try this out, but be prepared to follow up or back out if things break.
